### ISSUE TITLE
HEIST

### ISSUE THREAD

**Issue by igrigorik** [w3c/resource-timing#64]
[HEIST paper](http://papers.mathyvanhoef.com/blackhat2016.pdf), [ArsTechnica coverage](http://arstechnica.com/security/2016/08/new-attack-steals-ssns-e-mail-addresses-and-more-from-https-pages/), [Twitter discussion](https://twitter.com/bcrypt/status/760951421353152512)

Reading through the paper, the core observation is: TCP congestion control can be (ab)used to infer size of a (cross-origin) resource. Roughly:

- Identify when response headers are received:
    - [XHR readyState](https://xhr.spec.whatwg.org/#dom-xmlhttprequest-headers_received) (HEADERS_RECEIVED / 2) fires an event
    - Fetch [resolves](https://fetch.spec.whatwg.org/#fetch-method) the promise
    - Resource Timing exposes responseStart, subject to [timing allow check algorithm](https://w3c.github.io/resource-timing/#timing-allow-origin).
- Identify when the response body has finished loading
    - XHR readyState (DONE), onload() callbacks, Resource Timing’s responseEnd.

Given both of the above, you can compute the time delta between when the headers were received and response end:

- If delta is small then headers and response fit into same congestion window
- If delta is large then with high probability the response exceeded current congestion window (took multiple RTTs)

If TCP connection is brand new, the above can effectively tell you if response is <14KB. An extended version of this for beyond IW10 is:

- Open a connection and fetch a resource of known size to “pad” the congestion window with known amount of data.
- Request resource you want to know the size of and observe if it comes back in one or multiple RTTs - same criteria as above.

Aside, but related, it’s possible to [estimate RTT via javascript](https://www.smashingmagazine.com/2011/11/analyzing-network-characteristics-using-javascript-and-the-dom-part-1/): if you know the RTT and have a model for what you expect the congestion window to be, you can use  that observe if response took >1 RTT and arrive at same results with similar padding technique (knowledge of when response headers are received improves accuracy, but not necessary).

Armed with above, you can apply a compression oracle attack against the origin, ala Breach. That said, I’m dubious of the claims in the paper about how practical this actually is: tripping over a congestion window boundary doubles said congestion window and that ramps quickly and hence the query rate should be low... Am I missing something here?

Mitigations: all [existing BREACH recommendations](https://blog.qualys.com/ssllabs/2013/08/07/defending-against-the-breach-attack) apply.

---

In terms of practical implications... /cc @annevk @slightlyoff @domenic

- XHR / Fetch expose reponseStart for cross-origin resources; RT requires TAO.
    - Q: Anything we want or need to do here? 
- There are many ways to infer responseEnd..
    - Q: I’m not sure if and what can be practically done here?

Last but not least, scrubbing through RT spec it looks like we introduced a bug in https://github.com/w3c/resource-timing/commit/88bb58554c4fe792e5d4647544f5d2690498c5e1?diff=split#diff-eacf331f0ffc35d4b482f1d15a887d3bL543 ~implying that responseEnd is subject to TAO. It's not.. Unless @plehegar had something else in mind here?

**Comment by domenic** [comment]
I could use some help understanding how this attack works cross-origin. Shouldn't the fetch/XHR/resource timing/etc. automatically and instantly fail for such resources? (Assuming they are not shared via CORS.)

**Comment by djohns14** [comment]
CSP headers would be needed as they handle the outbound rules. CORS are inbound. GET and POST aren't blocked by CORS.

**Comment by annevk** [comment]
So one problem is that `fetch()` also resolves for "no-cors" when all headers are in. In combination with something else, e.g., `<img>`, you can then probably more accurately measure a response body. A mitigation would be to wait resolving `fetch()` for "no-cors" until the response body is fully transmitted.

**Comment by annevk** [comment]
I think https://github.com/whatwg/fetch/issues/355 is the solution here, though it would be interesting to hear from @tomvangoethem and Mathy Vanhoef why they think that is infeasible (per section 4.1.1 of their paper).

**Comment by annevk** [comment]
As @jakearchibald mentions that issue has obvious perf implications that are not desirable. Arguably resource timing should not expose responseEnd for no-cors, especially since it's not always readily available (it's not with `fetch()`). That still makes the platform susceptible to attacks that use multiple requests, one to figure out "start" and another for "end" (e.g., `<iframe>` load). However, that will get muddy with HTTP caches and such. We can also fiddle with when we expose "end" by delaying responses a little bit, but that would impact rendering.

None of this is ideal obviously.

**Comment by jakearchibald** [comment]
Would padding end impact render? The rendering of an iframe can't be detected. The width/height of an image can, but that's way before response end

**Comment by jakearchibald** [comment]
The problem is response timing is *designed* to give an accurate number here, and that's baaad. Others we can (hopefully) make less accurate.

**Comment by yoavweiss** [comment]
I'm wondering if we're not trying to route around the problem in hope that it would go away instead of tackling it. As mentioned in https://github.com/w3c/resource-timing/issues/64#issue-169492451, RTTs can be measured without any fancy APIs, using `<img onload>`.  Determining if a certain resource download took 1 or 2 RTTs doesn't require sub-millisecond precision in today's networks, unless the user is in the same data-center as the origin.

So, unless we're willing to delay certain events (and observable implications of resource loading) in tens of milliseconds to all users, this type of timing attack is not going away.

The underlying issue is that with BREACH, exposing response sizes equals exposure of CSRF tokens and login cookies. Are all types of compression vulnerable to BREACH? Can we devise a compression scheme that won't be? (e.g. by adding random padding chars of random size </naive-hand-waving>)

**Comment by domenic** [comment]
@yoavweiss can you produce a demo of this attack using only `<img onload>`, if that's all that's required? That would certainly make your point clearer.

**Comment by jakearchibald** [comment]
Delaying iframe load events seems ok, as long as it's fired after the frame has loaded.

Img is only useful if the response is a valid image, otherwise error can happen early.

**Comment by yoavweiss** [comment]
An image resource on the same host as the HTML under attack can be used to measure the RTT, as demonstrated in the [article](https://www.smashingmagazine.com/2011/11/analyzing-network-characteristics-using-javascript-and-the-dom-part-1/#measuring-network-latency-with-javascript) @igrigorik linked to. Since that article, img loading became async (in most browsers), but an attacker could probably use RAF in order to estimate the time gap between adding the img and when it's actually triggered. That code snippet above ignores potential connection establishment time which can skew its results, but that can be mitigated with `<link rel=preconnect>`, or by simply repeating an image fetch from the same host twice.

That would give an attacker one piece of the puzzle, which is the base RTT. And even if we'd delay `<img onload>`, this won't be mitigated until we would also delay the time in which the image request switches to the "partially available" state, since that time is highly observable, due to intrinsic dimensions changes. All image types but certain JPEGs are likely to switch to that state right when the first few bytes of image data get in.

@domenic - I hope that makes what I meant by "RTTs can be measured without any fancy APIs, using `<img onload>`" clearer.

Regarding delaying `<iframe>`'s onload, it might be helpful, but we'd need to delay it by a *lot* and we'd also need to delay onload for `<script>`, `<link>` and potentially others. Processing of the returned response on some of those might be observable, so we'd need to carefully figure that out and avoid such processing. 

@jakearchibald - would we also need to introduce similar delays to the time `fetch()` promises are resolved in SW?

**Comment by annevk** [comment]
RTT is not that useful on its own I think. What is important is time-to-headers, time-to-end-of-response-body, the difference between those, how much we expose in a single roundtrip, how much is exposed on several roundtrips, and how reliable those all are.

We can prevent exposing time-to-headers for no-cors, but the cost is a performance hit when you use service workers. (And an argument has been made that we also expose time-to-headers when you enable CORS, since CORS will fail at that point, but the request will also be different.)

We can also prevent exposing time-to-end-of-response-body for no-cors, at no cost, but then the question is whether you can still reliably determine it across several roundtrips and whether that is problematic.

**Comment by djohns14** [comment]
Excusing my ignorance in advance - 

If the purpose of resource timing is to provide performance metrics (effectively a reporting tool) to a web dev/web admin, why does the API have to return those statistics directly to the requester? Shouldn't the resource timing data be sent to a specified (by the page) directive, and not necessarily back to the requester?

**Comment by yoavweiss** [comment]
> RTT is not that useful on its own I think. What is important is time-to-headers

Good point that RTT and time-to-headers are only ~identical with static resources, where HEIST mostly targets dynamic components of a page (CSRF, cookies) which generation can add server-side time.


> We can also prevent exposing time-to-end-of-response-body for no-cors

Can you elaborate on that?

**Comment by annevk** [comment]
> Can you elaborate on that?

By not handing out responseEnd. That way if you get time-to-headers with `fetch()`, you don't get time-to-end-of-response-body. (You might still get that in some other way, but that would be a different request, and then the other variables come into play.)

**Comment by annevk** [comment]
@djohns14 I'm not sure what you're suggesting, but changing the fundamental nature of any API under discussion is not an option.

**Comment by jakearchibald** [comment]
I don't think a CORS request and a no-cors request are different enough to matter much, but maybe I'm missing something.

In terms of response end, we need to delay cache put, which is fine as it isn't all that perf sensitive. We might need to think about appcache writes too.

**Comment by sleevi** [comment]
From the peanut gallery, it does seem strongly inadvisable to try to just monkey-patch around this without articulating your threat model or assumptions. I think @yoavweiss has the right idea, which is pointing out we have a variety of leaks.

It's also important to keep in mind the context of what we're discussing, re: BREACH attacks, and the available server-side mitigations for these.

I'm not suggesting we "do nothing", but rather we work from first principles to make sure the platform is both secure and consistent. Minimally, that starts with elaborating the actual threat model we're wishing to defend against. Are we attempting to stop BREACH attacks? Are we attempting to stop knowledge of response body size? etc. Once we get those principles down, collectively take a look at the platform and look where we can leak that information, so we have a sense of the damage, and then we can brainstorm mitigations.

While this is very exciting, sexy, and arguably disturbing (in that "We probably should have seen this coming" feeling afforded by 20/20 hindsight), we shouldn't be reactionary. In my mind, this strikes as similar to concerns regarding privacy/tracking: concerns which are real, and grounded, but when you see something like https://www.chromium.org/Home/chromium-security/client-identification-mechanisms , you have a better appreciation for the holistic picture and where and why various piece-meal solutions ... aren't.

But that's just my $.02

**Comment by tomvangoethem** [comment]
The reason we didn't consider resolving `fetch()` until the complete body is in as a complete solution, was because if there is just a single observable event (either directly or indirectly) that happens when the headers of an opaque response are in, the attack resurfaces. While it may be possible to track down all the directly observable events, I am not convinced all possible side-channels can easily be identified.

Knowing the time-to-headers is indeed not a requirement. On connections where the jitter/RTT ratio is small, it shouldn't be too hard to determine whether 1 or 2 RTTs were needed (in contrast to 0 vs 1 in HEIST). As such, just knowing time-to-end-of-response-body will probably be enough.

Ideally, I would like to see a general defence where it's simply not possible to perform these attacks, regardless of the possible side-channels that may be exploited. Currently, the only way I see how this would work is by disabling authenticated `no-cors` requests (or simply stripping the cookies).

We are currently exploring a general technique based on leveraging the DEFLATE algorithm to counter BREACH-like attacks. However, simply knowing the length of a resource (regardless of compression) has its own security/privacy consequences (see https://github.com/whatwg/storage/issues/31 for example).

**Comment by jakearchibald** [comment]
If we rebooted the web today, any cross-origin communication would require CORS, or be no-credentials like you suggest.

I just don't see how we can do that now with 20 years of content depending on current behaviour.

**Comment by jakearchibald** [comment]
We could have a header to opt out of no-cors access, similar to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options

**Comment by jakearchibald** [comment]
I realise an opt-in isn't ideal, but some kind of `Access-Control-Allow-Credentials: NEVARR!` header would also mitigate https://github.com/whatwg/html/issues/1427

**Comment by annevk** [comment]
See my From-Origin header from some years back. We want something that prevents embedding from the wrong origin, including nested browsing contexts, but does not prevent top-level navigation. Note that omitting credentials does not help with intranet resources necessarily.

**Comment by jakearchibald** [comment]
https://www.w3.org/TR/from-origin/

Seems right, but applies to all requests, not just embedding (maybe this is the intent, the spec just talks about embedding a lot). What stopped work on this?

**Comment by annevk** [comment]
Yeah, it needs some tweaks, but we should not block top-level navigation. (And maybe we should continue allowing requests without credentials…)

Work stopped since nobodywanted to implement.

**Comment by jakearchibald** [comment]
> we should not block top-level navigation

`target="_blank"` should be ok too, since the client doesn't have access to a load event. `window.open` with `noopener` prevents load events too I think?

**Comment by annevk** [comment]
Yeah, probably. I'd like to sort out the whole popups a think a bit more and which browsing contexts end up being in a group together before firmly answering that. But definitely "noopener" has the semantics of creating a top-level rather than auxiliary browsing context.

So yeah, maybe reviving that header in some form in Fetch is a good step towards offering some kind of protection against this. Especially now https://w3c.github.io/webappsec-epr/ is being parked.

**Comment by igrigorik** [comment]
> Similar to the defense of blocking 3rd-party cookies, the web server could also block
requests that are not legitimate. A popular way of doing this, is by analysing the Origin
and/or Referer request headers. This allows the web server to determine whether the
request originated from its own domain or a potentially malicious one. However, it is
still possible to make requests without these two headers, preventing the web server to
determine where to request originated from. As a result, this technique can not be used
to prevent attacks.

@tomvangoethem can you elaborate on the cases where these headers are missing? As far as I can tell, if the server doesn't see a Referer header, or a Referrer header with an origin that it hasn't whitelisted, it can block the request and be done with it. What would that miss?

**Comment by djohns14** [comment]
@igrigorik great summary of the attack scope. The whitepaper, as you've pointed out, does theorize that HEIST could be used without Resource Timing. Resource Timing just makes the attack considerably easier. 

https://www.w3.org/TR/CSP3/

Ideally, anyone with a mission critical or otherwise sensitive web application enforces a Content Security Policy. Robust implementation of a CSP will stop HEIST in its tracks, regardless if Resource Timing is used or some other script. Unfortunately, a CSP white-list can quickly become unwieldy (albeit this has been considerably alleviated in CSP3). It may be prudent to encourage the use of a CSP in the Resource Timing specification. Interestingly enough, I saw no mention of CSP headers in the bh whitepaper. 

@annevk @jakearchibald I'm not sure what https://www.w3.org/TR/from-origin/ gets you over a CSP. frame-ancestors and plugin-types policy directives specifically address embedding. That being said, it certainly wouldn't make fetch() less secure.

> @djohns14 I'm not sure what you're suggesting, but changing the fundamental nature of any API   under discussion is not an option.

@annevk I was suggesting that access to all of this performance data (via the performance timeline) is somehow restricted to authenticated users and not just origins. 

As you know, there are two basic purposes of all the timing interfaces - (1) recording/measurement of entries (e.g. performance.now, mark) and (2) retrieval (e.g. getEntries, observe) of entries. I don't see any reason why access should be restricted for function (1). Function (2), however, provides the requester with some very detailed information and may warrant additional ways to ensure that only the developer has access to said information. This especially becomes true as more timing interfaces are created (e.g. Server Timing). 

I will not pretend to know how retrieval of performance metrics can be restricted, or if that constitutes a change which is fundamental in nature.  However, the same-origin policy is the only security policy in the current timing interfaces, which is woefully inadequate. That is not a knock on this working group. There just aren't many options for those who want to build security into their web applications (but there are some promising drafts). To that effect, it may be an opportune time to open a dialog with the web authentication working group and other security-focused W3 groups.

**Comment by jakearchibald** [comment]
@djohns14 for the most part, CSP is self-limiting. frame-ancestors is an exception, but we're talking about a rule that allows an origin to decide which origins can make no-cors requests to it, with the exception of top-level navigations which are always allowed.

CSP can't do this right now, and it certainly can't stop HEIST.

**Comment by tomvangoethem** [comment]
@igrigorik After top navigation from websiteA to websiteB, the `Referer` header will be set to websiteA, and no `Origin` header would be present. If websiteB were to block requests from unknown origins, then it simply wouldn't allow any visitor coming from external websites. The same idea was suggested as a defence for CSRF, but never got any traction (I don't know of any website that does this).

**Comment by igrigorik** [comment]
@djohns14 the attack is: malicious party X entices user to visit a site they control (to look at pictures of kittens, or some such), and while the user is busy admiring pretty pictures, the site is probing cross-origin resources to learn something about that user. For obvious reasons, malicious site X would not deploy CSP. Or, put differently, CSP is a great way for (good) sites to protect themselves from inadvertently becoming a launch point for such attacks (due to an XSS hole, or some such), but CSP does not address the case of a user visiting a malicious/compromised host. 

Re, access to perf data: use HTTPS, deploy CSP, resource integrity checks, etc. The origin controls what scripts it allows to execute, where those scripts are allowed to send data, and so forth. We don't need to invent anything new here.

@tomvangoethem ah, interesting. I'd love to see a matrix of all the current combinations of Origin/Referer headers across various request types.. I think that would help figure out how/what we can change to make this reasonable to deploy. @annevk are you aware of any such thing?

**Comment by jakearchibald** [comment]
https://tools.ietf.org/html/draft-west-first-party-cookies-07 was pointed out to me on Twitter. This would solve all but the intranet case.

**Comment by jakearchibald** [comment]
…except it also hides cookies for top level navigations. Haven't figured out why yet.

**Comment by jakearchibald** [comment]
We also need to ensure https://w3c.github.io/resource-hints/ doesn't leak load times. It specs load events pretty loosely, they probably shouldn't fire at all for cross-origin resources.

**Comment by sleevi** [comment]
On Monday, August 8, 2016, Jake Archibald <notifications@github.com> wrote:

> We also need to ensure https://w3c.github.io/resource-hints/ doesn't leak
> load times. It specs load events pretty loosely, they probably shouldn't
> fire at all for cross-origin resources.
>

I still suggest you're modeling it wrong if you're thinking about it like
that. Think about the threat model first before you try to go plug timing
left and right.

Fundamentally, this is a server issue.

**Comment by jakearchibald** [comment]
@sleevi I'm thinking the best way we can do this is to provide an opt-out for cross-origin no-cors requests for everything except top level navigations.

Samesite cookies may be the answer.

I bring up prerender because it's a bit ambiguous on whether it's a top-level navigation or not, so it's an exceptional case we need to consider.

Or am I still barking up the wrong tree?

**Comment by annevk** [comment]
@igrigorik no, there's in fact numerous issues around `Origin` and I believe only Chrome implements it as some XSRF mitigation for `<form>` POSTs. (I think I'd prefer `Origin` as an exclusive CORS header.)

@jakearchibald the reason you need to hide cookies for top-level navigation is that otherwise you can craft a URL that uses an open redirector or some such and does exfiltration at the same time. Of course, that makes it useless for login.

It seems very hard to protect against malicious requests. Even with https://github.com/mikewest/origin-policy we can only do it post-first-request. Only taking action if a request includes some non-browser-provided token is probably best. Not sure if the browser can help here.

That still doesn't help with timing attacks against requests, whereby you do a request without and with credentials and observe the difference. Same-site cookies can help here, maybe, but since they're not useful for login they wouldn't protect the important bits.

For responses we can do something like `From-Origin` and I think we should. We shouldn't prevent direct top-level navigation, but everything else seems okay. It will harm embedding, but it will also make it easier for servers to protect the user's identity and preferences.

**Comment by tomvangoethem** [comment]
@annevk Does returning a network error in `From-Origin` mean that the TCP connection is killed? If not, having it as a header won't properly protect against HEIST (and probably related issues either).
For example, consider following resources on a web server, with their size in number of TCP windows: `/secret?x=a` → 7 TCP windows, `/secret?x=b` → 8 TCP windows, `/img` → 17 TCP windows. Let's also assume that there's a `From-Origin` header on all resources except for `/img` (of course this could be any resource).
Now the attack looks as follows: attacker forces request to `/secret?x=a`, but can't determine size because of the network error. Using the same TCP connection, he forces a request to `/img`, of which he can determine the number of TCP windows (as it doesn't have the header). In this case, it will be a single TCP window (7 acknowledged packets made the congestion window grow from 10 to 17). Then the attacker does the same for `/secret?x=b` and notices that `/img` requires 2 TCP windows.
If this (or a similar) approach is used, it might be useful to consider delivering this information through some type of global manifest (this seems pretty hard to do with dynamic pages though...)

**Comment by annevk** [comment]
@tomvangoethem good point, so perhaps `From-Origin` only makes sense as part of @mikewest's Origin Policy work (I referenced it above). So either you protect all resources in an origin, or none.

**Comment by jakearchibald** [comment]
Killing the connections after a `From-Origin` failure seems reasonable. It's the exceptional case.

**Comment by igrigorik** [comment]
Before we rush ahead and spec more bandaid solutions, can we take a step back and understand why the current solutions are not sufficient or have not been successful? 

### Origin
According to the RFC, `Origin` was meant to be [sent on every request](https://tools.ietf.org/id/draft-abarth-origin-03.html#ua-behavior), not just on CORS preflights. Further, UAs were supposed to advertise [`null` token](https://wiki.mozilla.org/Security/Origin#Selection_of_.22null.22_token) in [variety of cases](https://wiki.mozilla.org/Security/Origin#When_Origin_is_served_.28and_when_it_is_.22null.22.29) to indicate that "cause of the request is not trustworthy" and as a signal that UA supports `Origin`. Except, based on my cursory examination.. it seems like UA's decided to omit `null` case. 

> While all cross-origin requests will contain an Origin header, some same-origin requests might have one as well. For example, Firefox doesn't include an Origin header on same-origin requests. But Chrome and Safari include an Origin header on same-origin POST/PUT/DELETE requests (same-origin GET requests will not have an Origin header). - [html5rocks](http://www.html5rocks.com/en/tutorials/cors/#toc-adding-cors-support-to-the-server)

So, developers think we send Origin all requests, but we don't.. Confusion abounds. Questions:

1. Why was the `null` case omitted by UA's? 
2. If we were to send `null` would it help? If not, why not?

Further, there is some confusion here with `Referrer` which I don't fully understand yet. At a minimum, it seems that [Referrer policy](https://www.w3.org/TR/referrer-policy/) should somehow affect what's advertised by `Origin`.

### From-Origin
The intent here seems to be for the origin to [return a whitelist](https://www.w3.org/TR/from-origin/), which the UA checks once it has the response and acts on (e.g. errors out the response)...

1. The client must fetch the response, which itself is subject to timing attacks
2. The server doesn't know if client will actually enforce said rules and thus must return the authenticated response and hope that the client enforces its whitelist. But, even if client fails the request, there is still the timing vector - see (1). 

In short, I don't see how https://github.com/whatwg/fetch/issues/365 would solve much, as it's still susceptible to timing attacks and provides rather weak security guarantees and benefits. 

### Same-Site cookies

[Same-Site cookies](https://tools.ietf.org/html/draft-west-first-party-cookies-07) goes a long way towards addressing the type of attacks we're discussing here. 
 - "Strict" mode it provides strong protection, modulo [top-level navigation considerations](https://tools.ietf.org/html/draft-west-first-party-cookies-07#section-5.2).
 - "Lax" provides [reasonable protection](https://tools.ietf.org/html/draft-west-first-party-cookies-07#section-4.1.1), except for top-level navigation cases.

Either way, encouraging UA and developer adoption of this seems like a win all around. And speaking of which, what is the implementation status of this proposal?

---

Stepping back, if we want to reduce the attack surface for the types of timing/size attacks we're discussing here, we can either block requests before they go out, or give the server sufficient information on the request such that it can alter the response. Client-side enforcement on the response doesn't help.

Long story short, exactly what @sleevi said: **"Fundamentally, this is a server issue."**

**Comment by annevk** [comment]
### Origin

We could not send `Origin` for every request because @abarth discovered that was not web-compatible. Note that https://github.com/whatwg/fetch/issues/225 is the open issue on defining the semantics of the `Origin` header for non-CORS cases. I've been waiting for input from Google for a while now.

The confusion with `Referrer` is that we cannot let CORS be influenced by the referrer policy. That would just break things at this point. We could let the non-CORS case be influenced I suppose. That should not break much since that has not been interoperable.

### From-Origin

I agree that From-Origin would mostly be defense-in-depth at first, and useful once all user agents support it. It's basically the counterpart to CORS and unlike same-site cookies helps with HTTP authentication, TLS client certs, and intranet scenarios. 

If From-Origin was deployed origin-wide, you could also make it so that only the first request can be attacked. Basically, with @mikewest's Origin Policy work, I think we can do a lot of things similar to HSTS (TOFU). And then maybe we can start preloading... And then maybe once we have enough static guarantees of sites being secure, we can offer them more powerful APIs. One can dream.

### Same-Site

I think Chrome might be shipping and Firefox plans on implementing though I can't find a bug.

**Comment by jakearchibald** [comment]
If request headers let a server know:

* The initiating origin, even if same-origin, or `null` if referrer information is not available/blocked
* If the request is a sub-resource, auxiliary navigation (anything where the initiating origin gets a `WindowProxy` and/or a load event), top level navigation

…then it could implement either From-Origin or same-site itself. This would be a big change for `Origin`, as all the NOs on https://wiki.mozilla.org/Security/Origin#When_Origin_is_served_.28and_when_it_is_.22null.22.29 would need to become YESs. `Origin2` anyone?

**Comment by annevk** [comment]
That wiki page does not seem accurate (definitely not for same-origin and also gives the idea "null" is transmitted which it's not as far as I know, except for CORS).

**Comment by annevk** [comment]
I suspect `Origin` could not be included everywhere due to CORS. So maybe if we want to go down that path we should restrict `Origin` to cross-origin "cors" requests and use `Bikeshed` for all other requests. `Bikeshed` would simply use `null` per referrer policy and such (although I suspect folks might want flexibility there too, have `Bikeshed` expose same-origin, but `Referer` omitted).

**Comment by jakearchibald** [comment]
With `Bikeshed` the cookies are still sent, so that might reveal something in the timing. Not sure how much that matters, but same-site has it covered.

**Comment by annevk** [comment]
We basically need to decide if these matter:

* HTTP authentication
* TLS client certificates (and maybe the various others mentioned in https://github.com/whatwg/fetch/issues/341#issuecomment-234977748)
* Firewalled content

Then, we need to decide what `Origin` should cover:

* CORS (not really negotiable, only cross-origin or cross-origin to same-origin)
* The extras that Chrome/WebKit implement (all non-HEAD-non-GET requests, including same-origin?). Also, does that help enough or we getting sites to pay attention to the wrong thing?

I don't think we need to decide whether same-site cookies should happen. It seems they're an obvious win.

Based on what we decide for `Origin` we can evaluate `Bikeshed`. And for `From-Origin` I think the only interesting variant is something that's origin-wide and installed upon first visit or even preloaded. That can really help isolate yourself from attackers.

**Comment by jakearchibald** [comment]
@annevk 

> We could not send Origin for every request because @abarth discovered that was not web-compatible

Any details on this?

**Comment by annevk** [comment]
@jakearchibald no, but see my speculation above how it's likely due to CORS.

**Comment by igrigorik** [comment]
_(side note: Adam pointed out that I was linking to an outdated draft earlier in the thread, we should be referring to [The HTTP Origin Header Field](https://tools.ietf.org/html/rfc6454#section-7) in RFC6454 instead)._

I asked Adam about the change from "MUST send" in earlier drafts to "MAY send" in the RFC: "the IETF didn't want to publish a spec that made every existing HTTP implementation non-conformant". That said, note that UA requirements section leaves this wide open:

> The user agent MAY include an Origin header field in any HTTP request.
> Whenever a user agent issues an HTTP request from a "privacy-sensitive" 
> context, the user agent MUST send the value "null" in the Origin header field.
> 
>  NOTE: This document does not define the notion of a privacy-
>  sensitive context.  Applications that generate HTTP requests can
>  designate contexts as privacy-sensitive to impose restrictions on
>  how user agents generate Origin header fields.

So, change from MUST to MAY is/was a spec-compat issue. The UA is free to send `Origin` on any request, and it seems that current implementations simply chose to limit themselves to CORS and POST requests. Which, given what we know today, is a decision that we should probably revisit...

- ~Send `Origin` on any cross-origin request (cors and no-cors)?
- ~Define some interop with Referrer. E.g. if Referrer policy is set to none, then send "null" in Origin header, or some such.

**Comment by jakearchibald** [comment]
> Send Origin on any cross-origin request

It would need to be all requests right? If I'm trying to emulate `From-Origin`, I only want to respond if `Origin` is my origin. Still need a header that indicates top-level vs aux.

**Comment by annevk** [comment]
Again, I don't think we can send `Origin` on any request and I'm pretty sure Adam found that out too when he tried back in the day. Especially now CORS is much more in use it's highly likely servers use it as conditional and end up breaking navigation and such.

**Comment by igrigorik** [comment]
@jakearchibald if all UA's sent `Origin` on every cross-origin request (cors and no-cors) then the absence of such header would allow the server to restrict responses to "my origin only", as well as whitelist particular origins. That said, since we already have Origin implementations out in the wild that don't conform to this... it makes it complicated to deploy. With that in mind: 

1. @annevk, your `Origin` proposal in https://github.com/whatwg/fetch/issues/225#issuecomment-238874536 makes sense to me. 
1. By the looks of it, we need new `Bikeshed` header (le sigh.. but at least now we know what we need to solve for :)) that would cover the no-cors cases, top-level vs aux, etc.

**Comment by jakearchibald** [comment]
@igrigorik good point. I was thinking no-referrer would also remove the origin header, but it would make it explicitly null. Happy for `Bikeshed` to follow these rules.

So are you suggesting that `Bikeshed` would also include the top-level/aux info rather than a seperate header? I guess that's fine, just needs more parsing.

One less header to vary on.

**Comment by igrigorik** [comment]
Tried to write down what we could recommend to developers with existing mechanisms...

---

Use First Party (FP) cookies for all authenticated content that's not meant to be embedded or accessible cross-origin: "Strict" mode provides strong protection, "Lax" provides reasonable protection (modulo top-level navigations). With FP in place...

If `Origin` header is present in the request, then it's either a CORS request, or a request whose method is neither HEAD or GET (https://github.com/whatwg/fetch/issues/225#issuecomment-238874536). Inspect the value of Origin header for the source origin of such requests to determine if you want to allow such request - see [HTTP access control (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Overview).

Otherwise, if `Origin` request header is missing, the request is either:
- same-origin, in which case the FP cookie should be present. 
- cross-origin, in which case the FP cookie is subject to policy and initiator:
 - If FP policy is "Strict", then the cookie should not be there.
 - If FP cookie policy is "Lax", then FP cookie is only present for top-level navigation requests that use a "safe" HTTP method.
    - When using "Lax" FP policy you can distinguish same-origin requests and cross-origin top-level navigation requests by inspecting the `Referrer` header: as long as your policy is not set to [no-referrer](https://www.w3.org/TR/referrer-policy/#referrer-policy-no-referrer), same-origin requests will contain a `Referrer` header whose origin is your origin.

---

BigBank.com has authenticated resources that should not be accessible cross-origin. It can either:
 1. set FP cookie policy to "Strict" and be done with it.
 2. set FP cookie policy to "Lax" and use combination of Origin and Referrer headers to distinguish between same-origin requests and cross-origin top-level nav requests, and potentially disallow the latter.

---

Do we need anything else? It seems like combination of FP, Origin, and Referrer might do the trick.

**Comment by annevk** [comment]
We can only give this advice if a) all browsers implement same-site cookies, b) all browsers implement the "new" `Origin` header semantics, c) we couple it with an HTTPS requirement.

And again, as I said before, this does not deal with 1) HTTP authentication, 2) TLS client certificates, 3) firewalled content.

**Comment by annevk** [comment]
If we did `Bikeshed` by the way, I think b) would be better addressed by all browsers implementing `Bikeshed` and only using `Origin` for CORS as that would improve the CORS protocol.

**Comment by igrigorik** [comment]
Contrary to what many may believe, I do prefer solutions that avoid minting new headers. :bowtie: Especially so when we're talking about a header that would have to be attached on most every request—which, I think, is what were looking at here. Hence, me trying to understand if we actually _need_ a new header, of if we can compose a solution out of existing parts. 

> We can only give this advice if a) all browsers implement same-site cookies, b) all browsers implement the "new" Origin header semantics, c) we couple it with an HTTPS requirement.

As far as I can tell, Chrome is already effectively (b), we have (a) on the way, and in this day and age (c) goes without saying... 😎 . Further, both (a) and (b) have existing efforts behind them, so if we can give them a kick through additional use-cases/motivation, then that's a win in my books. Building our own thing will take just as long if not longer.

That aside, you're right, I ignored HTTP/TLS auth and firewalled content. It wasn't clear to me how it impacts what we're discussing here.. can you elaborate a bit more?

**Comment by annevk** [comment]
You could still do timing attacks on resources that are firewalled or HTTP/TLS authenticated. Basically the same as with cookies.

(It's not entirely clear by the way what `Origin` adds over just using lax cookies.)

**Comment by igrigorik** [comment]
> You could still do timing attacks on resources that are firewalled or HTTP/TLS authenticated. 

Hmm, well.. you could just drop a first-party cookie on any such resources, right? Same logic.

> It's not entirely clear by the way what Origin adds over just using lax cookies.

Yeah, that's fair, you can probably simplify my earlier post to FP + Referrer. 

---

Stepping back, we covered a lot of ground in this thread.. My takeaways: 
- we are *not* making any changes to Fetch or Resource Timing
- we _should_ encourage developers to deploy server-side logic to protect sensitive resources
 - BREACH recommendations apply but we established that `Referrer` alone is not sufficient
 - First-Party cookies goes a long way towards resolving this problem:
   - In strict mode, it provides strong protection.
    - in lax mode, it can be combined with Referrer to provide strong protection.

With above in mind, I propose that we resolve this thread and go nudge folks working on FP. I can writeup a summary of our discussion here.

**Comment by jakearchibald** [comment]
> in lax mode, it can be combined with Referrer to provide strong protection.

It's still not clear to me when cookies would be sent with "lax" mode. What about:

* `<a href="…">` - yes
* `<iframe>` - no
* `window.open` - I would hope no
* `<a href="…" target="_blank">` - I would hope yes
* `<link rel="prerender">` - ??? (hence https://github.com/w3c/resource-hints/issues/63)

**Comment by igrigorik** [comment]
> It's still not clear to me when cookies would be sent with "lax" mode.

The behavior is defined in https://tools.ietf.org/html/draft-west-first-party-cookies-07#section-4.3. If that's not sufficient, then we should open a bug against httpwg: https://github.com/httpwg/http-extensions/issues?q=is%3Aopen+is%3Aissue+label%3A6265bis /cc @mikewest @mnot 

Re, prerender: I'll followup on the linked issue.

**Comment by igrigorik** [comment]
Updated RT spec (7358cbd) and removed the (broken) TAO-check reference -- see my very first post + linked commit on this thread for details. I believe that's the only actionable change for RT from this discussion.. If anyone else has other suggestions, let me know.

**Comment by jakearchibald** [comment]
Seems like `<a href="…" target="_blank">` does not create a top-level browsing context. That's not so great, I'd expect to get cookies in "lax" mode in that case.

`window.open` won't get cookies, which is great (although we may want to allow them with the `noopener` option. `clients.openWindow()` is top level, so it will get cookies.

**Comment by jakearchibald** [comment]
Filed https://github.com/httpwg/http-extensions/issues/226

**Comment by annevk** [comment]
> Hmm, well.. you could just drop a first-party cookie on any such resources, right? Same logic.

I don't understand.

**Comment by mikewest** [comment]
> Use First Party (FP) ...

Nit: We renamed this before shipping, so if/when you write anything up, please try to minimize the confusion. :)

> <a href="…"> - yes
> <iframe> - no
> window.open() - I would hope no
> <a href="…" target="_blank"> - I would hope yes
> clients.openWindow() - I would hope yes
> <link rel="prerender"> - ??? (hence w3c/resource-hints#63)

As noted in the bug Jake filed against SameSite, `Lax` same-site cookies would be sent with all of these (presumably cross-origin) (pre-)navigations, with the exception of `<iframe>` (because auxiliary browsing contexts are also top-level browsing contexts). That was intentional, as `SameSite`'s target was CSRF, so the request is what mattered for the cookie state, not the exposure via `Window` objects.

**Comment by igrigorik** [comment]
@annevk as in, regardless of what authentication mechanism you use (HTTP auth, TLS auth), when you first authorize the user you can just drop a SameSite cookie and then observe if it's echoed in subsequent requests to identify if its origin is same or cross origin.

**Comment by annevk** [comment]
I'm starting to agree with @mikewest's remarks elsewhere that the opt-in defense is not ideal and unlikely to be deployed. But I'm not sure how we can do better, other than maybe making the opt-in not rely on new cookie infrastructure, but something closer to a boolean.

**Comment by sleevi** [comment]
While I'm sympathetic to @mikewest's position, I'm absolutely convinced this is not something that the browser caused, contributes to, or can fix at this point. This is, at the core, similar to something like SQL injection - it's a server choice (to compress responses), and using compression - whether over a secure channel or otherwise - leaks status information.

While I appreciate @igrigorik's efforts at coming up with a threat model, I think it's utterly futile to suggest that the browser could be in a position so it could make all loads unobservably side-effect free, without the cooperation or knowledge of the server. There are going to be timing leaks throughout the system - whether at the CPU cache layer, at the IO interrupt layer from the NIC, from contention on the network, from congestion windows, from system timers - all of these things outside of the browsers' ken and remit. We simply can't design a system that, under this threat model, is constant-time, and without that aspect of constant-time, we cannot guarantee that the secrets remain secret.

So we have to do one of two things - prevent requests from being made that might result in secrets being sent, or help server operators understand the risks of sending secrets. While SameOrigin or Bikeshed or whatever we want can quais-help with the former, the past two decades of the Web have *also* taught us that servers are, almost universally, very bad at determining what is or should be secret (case study: The opposition to HTTPS). So even if we were to prevent most (all?) forms of cross-origin credential sharing, there's going to be secrets, and so there's going to be side-channels here. Which leads to the only other option that seems at all practical, which is educating at a server side.

I think the focus on the "Web Platform Features" enabling this are arguably misguided, because what we're talking about is the ability for constant-time vs variable-time, and anyone who works in that space can tell you how hard it is to make sure it's right. 

To the extent we could block cookies (and TLS client certs, and whatever H/2+TLS1.3 method the IETF comes up with), great, but I think we know that the risk of breakage is high for any solution, because it's not backwards compatible, and thus our option is "Opt In". Or tell servers to stop supporting compression if they can't do it securely. Or stop advertising in browsers that we support compression.

**Comment by igrigorik** [comment]
> I think it's utterly futile to suggest that the browser could be in a position so it could make all loads unobservably side-effect free, without the cooperation or knowledge of the server. So we have to do one of two things - prevent requests from being made that might result in secrets being sent, or help server operators understand the risks of sending secrets.

Agreed, and my direction here has been to figure out what (if anything) we need to give server operators to make it possible to make an informed decision on their end for whether the response should be allowed to contain secret / authenticated data.

> I think the focus on the "Web Platform Features" enabling this are arguably misguided, because what we're talking about is the ability for constant-time vs variable-time, and anyone who works in that space can tell you how hard it is to make sure it's right.

So, to clarify.. The problem, as I see it, is that we allow authenticated cross-origin requests that the origin server can't distinguish from same-origin requests. As a result, the server ends up leaking secrets because existing recommendations (e.g. look at `Referrer` / `Origin` headers) would also block top-level navigations. So, we do need some new "web platform feature" to allow servers to solve this.. SameOrigin cookies might be sufficient, or maybe we need Bikeshed instead.

**Comment by igrigorik** [comment]
A quick summary of our discussion here: https://www.igvita.com/2016/08/26/stop-cross-site-timing-attacks-with-samesite-cookies/ - tl;dr: use SameSite cookies.

As outlined in https://github.com/w3c/resource-timing/issues/64#issuecomment-239863341, we are not making any changes to RT or Fetch; closing this thread. Perhaps there are other mechanisms we ought to consider, in addition to encouraging adoption of SameSite cookies, but whatever that may be.. we should take that discussion to the appropriate forum (webappsec group, probably).

_p.s. thanks everyone for your help and input!_
